core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java

```diff
@@ -295,30 +234,106 @@ public boolean equals(Object obj) {
+            public CompoundOrderComparator(List<BucketOrder<B>> compoundOrder, Aggregator aggregator) {
                 this.compoundOrder = compoundOrder;
                 this.aggregator = aggregator;
             }
 
             @Override
-            public int compare(Bucket o1, Bucket o2) {
+            public int compare(B o1, B o2) {
                 int result = 0;
-                for (Iterator<Terms.Order> itr = compoundOrder.iterator(); itr.hasNext() && result == 0;) {
+                for (Iterator<BucketOrder<B>> itr = compoundOrder.iterator(); itr.hasNext() && result == 0;) {
                     result = itr.next().comparator(aggregator).compare(o1, o2);
                 }
                 return result;
             }
         }
     }
 
+
+    private static final byte COUNT_DESC_ID = 1;
+    private static final byte COUNT_ASC_ID = 2;
+    private static final byte KEY_DESC_ID = 3;
+    private static final byte KEY_ASC_ID = 4;
+
+    /**
+     * Order by the (higher) count of each term.
+     */
+    public static final InternalOrder COUNT_DESC = new InternalOrder<>(COUNT_DESC_ID, "_count", false, Comparator.comparingLong(Bucket::getDocCount).reversed());
+
+    /**
+     * Order by the (lower) count of each term.
+     */
+    public static final InternalOrder COUNT_ASC = new InternalOrder<>(COUNT_ASC_ID, "_count", true, Comparator.comparingLong(Bucket::getDocCount));
+
+    /**
+     * Order by the terms.
+     */
+    public static final InternalOrder KEY_DESC = new InternalOrder<>(KEY_DESC_ID, "_key", false, comparingKeys().reversed());
+
+    /**
+     * Order by the terms.
+     */
+    public static final InternalOrder KEY_ASC = new InternalOrder(KEY_ASC_ID, "_key", true, comparingKeys());
+
+    private static int compareKey(Histogram.Bucket b1, Histogram.Bucket b2) {
+        if (b1 instanceof InternalHistogram.Bucket) {
+            return Double.compare(((InternalHistogram.Bucket) b1).key, ((InternalHistogram.Bucket) b2).key);
+        } else if (b1 instanceof InternalDateHistogram.Bucket) {
+            return Long.compare(((InternalDateHistogram.Bucket) b1).key, ((InternalDateHistogram.Bucket) b2).key);
+        } else {
+            throw new IllegalStateException("Unexpected impl: " + b1.getClass());
+        }
+    }
+
+    private static Comparator comparingKeys() {
+        return (o1, o2) -> {
+            if (o1 instanceof Terms.Bucket && o2 instanceof Terms.Bucket) {
+                return ((Terms.Bucket) o2).compareTerm((Terms.Bucket) o1);
+            }
+            if (o1 instanceof Histogram.Bucket && o2 instanceof Histogram.Bucket) {
+                return compareKey((Histogram.Bucket) o1, (Histogram.Bucket) o2);
+            }
+            throw new UnsupportedOperationException("order not support for buckets " + o1.getClass().getName() + " and " + o2.getClass().getName());
+        };
+    }
+
+    public static boolean isCountDesc(BucketOrder order) {
```

Comment:
Not sure what you mean here? Do you mean it will return false if its passed a `Compound` order with only 1 element which is `InternalOrder.COUNT_DESC`?